[1.12.x] Implement hooks allowing mods to add horse armor. Partially addresses #3975.

Daomephsta opened the pull request:
**What does this PR do?**
It allows mods to add their own horse armor. To add a new horse armor item, `Item#getHorseArmorType()` is overridden  to returns a `HorseArmorType` enum constant other than `HorseArmorType.NONE`. New enum constants can be added to `HorseArmorType` using `EnumHelper.addHorseArmor()`. Code can be run when the horse wearing the armor updates by overriding `Item#onHorseArmorTick()`.

**Testing**
A test mod (HorseArmorTest) is included in this PR. It is disabled by default. It adds a new `HorseArmorType` with a strength of 15, and an item that uses that `HorseArmorType`. The item also applies the glowing effect to the horse wearing it, via `Item#onHorseArmorTick()`. It will change texture to iron armour if enchanted. 

Forge-Vanilla compatibility is not broken by this PR. I tested:  

- Vanilla server with two Forge clients
- Vanilla server with one Forge client and one Vanilla client
- Forge server with two Forge clients
- Forge server with one Forge client and one Vanilla client
- Forge server with two Vanilla clients  

In all these situations all clients were able to see changes in the equipped horse armour.
I have also tested that removing a horse armor adding mod will not crash the world if a horse was wearing the added armor. 

**Other**
Parts of this PR were based on #3185. I have done my best to fix the issues mentioned in that PR. Methods and constructors added by this PR are documented where necessary and old methods and constructors are deprecated. Load order should not cause issues as the system is based on the name of the enum constant, rather than the ordinal.
Partially addresses #3975.

leviathan143 committed:
Implement hooks allowing mods to add horse armor

leviathan143 committed:
Add vanilla armor to Item#getHorseArmorType()

leviathan143 committed:
Move the call to Item#onHorseArmorTick to a place where it will actually be called every tick.

Add a test case for Item#onHorseArmorTick to the test mod.

leviathan143 committed:
Convert tabs to spaces

leviathan143 committed:
Remove deprecated methods

Daomephsta commented:
The EnumHelperTest test is failing because there aren't accessors for the ctors with the parameters (int armorStrengthIn, String p_i46800_4_, String p_i46800_5_) & (int armorStrengthIn). Neither of these ctors should be accessible through EnumHelper (The former forces textures to be in the vanilla domain and in a specific folder, the latter only exists to construct HorseArmorType.NONE). What should I do?

leviathan143 committed:
Merge branch '1.12.x' into horse-armour-1.12

Daomephsta commented:
@mezz Now that the 1.12 dust has settled, can I get an update on this PR's status? I know I need to revert 7a26fde as the breaking changes window has closed, are there any other issues?

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,71 @@
++            return p_188580_0_.func_77973_b().isValidHorseArmor(p_188580_0_) ? p_188580_0_.func_77973_b().getHorseArmorType(p_188580_0_) : NONE;
+     }
+-
+-    public static boolean func_188577_b(Item p_188577_0_)
```
Don't remove item aware versions of methods. Deprecate them instead.

liach commented on patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:
```diff
@@ -0,0 +1,63 @@
+     private static final UUID field_184786_bD = UUID.fromString("556E1665-8B10-40C8-8F9D-CF9B1667F295");
+     private static final DataParameter<Integer> field_184789_bG = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
+     private static final DataParameter<Integer> field_184791_bI = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
++    private static final DataParameter<String> HORSE_ARMOR_STRING = EntityDataManager.<String>func_187226_a(EntityHorse.class, DataSerializers.field_187194_d);
```
You can just use the int one as long as there's no stupid modder adding horse armor after the completion of initialization.

Daomephsta replied:
It's string based because that prevents load order changes causing strange behaviour. The vanilla system is ordinal based. Mod removal, addition or dependency declaration could all cause the ordinal of a non-vanilla enum constant to change, which would result in the wrong armour appearing on a horse. I could add an ID field to HorseArmorType, but then ID collisions are more likely.

LexManos replied:
Ya, Honestly tho, it may be worth it just to send the full ItemStack instead of just a string. As modders may make textures determinent on NBT.

Daomephsta replied:
It doesn't look like that would be too hard to do. I'll give it a try.

liach replied:
@Leviathan143 Can you fix this then?

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -714,6 +714,32 @@
++     * @param horse - The horse wearing the armor
++     * @param itemStack - the armor itemstack
++     */
++    public void onHorseArmorTick(World world, net.minecraft.entity.passive.EntityHorse horse, ItemStack itemStack) {}
```
I'd suggest you moving these methods to a capability or an interface.

Daomephsta replied:
As far as I'm aware, Forge discourages that. I think this did change with 1.12 though, can you confirm that?

liach replied:
This method is really rarely used...

liach commented on src/test/java/net/minecraftforge/debug/HorseArmorTest.java:
```diff
@@ -0,0 +1,84 @@
+            testArmorType = EnumHelper.addHorseArmor("test", MODID + ":textures/entity/horse/armor/test.png", "tst", 15);
+            testArmorItem.setRegistryName(MODID, "test_armor");
+            testArmorItem.setUnlocalizedName(MODID + ".testArmor");
+            GameRegistry.register(testArmorItem);
```
Use `ForgeRegistries.ITEM.register()` instead

williewillus replied:
it should really be using the registry events

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -714,6 +714,32 @@
++     * @param stack - The armor stack
++     * @return True if the itemstack is valid horse armor, false if it isn't.
++     */
++    public boolean isValidHorseArmor(ItemStack stack)
```
Can be combined with `getHorseArmorType`

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -714,6 +714,32 @@
++     * @return The HorseArmorType of the armor stack 
++     */
++    @Nullable
++    public net.minecraft.entity.passive.HorseArmorType getHorseArmorType(ItemStack stack)
```
Make it return `NONE` instead of `null`. Remove `@Nullable`.

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,71 @@
++        if(p_188580_0_.func_190926_b())
+         {
++            return NONE;
++        }
```
Just call `getHorseArmorType(stack)` instead of dicking with vanilla code.

liach commented on patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:
```diff
@@ -0,0 +1,63 @@
+             this.field_70180_af.func_187230_e();
+             this.func_110230_cF();
+         }
++        this.field_110296_bG.func_70301_a(1).func_77973_b().onHorseArmorTick(field_70170_p, this, this.field_110296_bG.func_70301_a(1));
```
Should cache the horse armor item stack

leviathan143 committed:
Revert "Remove deprecated methods"

This reverts commit 7a26fde4e556e444400eac05d45e918cb71eae0a.

leviathan143 committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into horse-armour-1.12

leviathan143 committed:
Fix most raised issues

leviathan143 committed:
Merge branch 'horse-armour-1.12' of https://github.com/Leviathan143/MinecraftForge into horse-armour-1.12

leviathan143 committed:
Resolve merge conflicts

leviathan143 committed:
Add acceptableRemoteVersions = "*" so that vanilla clients can connect to forge servers with this mod

leviathan143 committed:
Convert tabs to spaces

leviathan143 committed:
Move horse armor hooks to an interface, IHorseArmor

Daomephsta commented:
Right that's all the issues addressed. However I still don't know what to do about the EnumHelperTest issue

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -34,14 +34,18 @@
      {
 -        return p_188580_0_.func_190926_b() ? NONE : func_188576_a(p_188580_0_.func_77973_b());
-+        return p_188580_0_.func_77973_b().getHorseArmorType(p_188580_0_);
++        if(p_188580_0_.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor)
```
Should this be an interface or a capability?

covers1624 replied:
Interface, makes absolutely no sense to be a cap and nothing will be gained from it being a cap.

covers1624 commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,20 @@
+     * @param stack - The armor stack
+     * @return The HorseArmorType of the armor stack 
+     */
+    public net.minecraft.entity.passive.HorseArmorType getHorseArmorType(ItemStack stack);
```
Why not an import? This is a forge class.

liach replied:
Because this is copied from patches.

covers1624 commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,20 @@
+     * @param horse - The horse wearing the armor
+     * @param itemStack - the armor itemstack
+     */
+    public void onHorseArmorTick(World world, net.minecraft.entity.passive.EntityHorse horse, ItemStack itemStack);
```
Same here.

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,62 @@
+     }
+ 
++    //Allows for textures located outside the vanilla horse armor folder
++    private  HorseArmorType(String textureLocation, String hash, int armorStrengthIn) 
```
two spaces betweeen `private` and `HorseArmorType`

liach replied:
And no need for `private`

Daomephsta replied:
It's an enum constructor. Why shouldn't it be private?

liach replied:
It can be package private based on the fact that only enum constants can use the constructor without reflection or asm

Daomephsta replied:
Ok, but what's wrong with private? No one has any business accessing that ctor except through the EnumHelper method. Is there a Java convention or Forge convention to use the highest access level you can use without unnecessarily exposing classes/fields/methods?

liach replied:
Then fix the space :cry:

liach commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,19 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
+    @Nullable 
+    public static HorseArmorType addHorseArmor(String name, String textureLocation, String hash, int armorStrength)
+    {
+        return addEnum(HorseArmorType.class, name, textureLocation, hash, armorStrength);
```
This one is wrong... You don't need a `name` argument and you shouldn't have one. That's why the test is failing.

liach commented:
You probably need to do a special check like this for the failing test: https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/test/java/net/minecraftforge/test/EnumHelperTest.java#L89

leviathan143 committed:
Exclude internal constructors of HorseArmorType from being checked by EnumHelperTest

leviathan143 committed:
Use imports instead of fully qualified names in Forge classes

leviathan143 committed:
Fix double space

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,62 @@
+     }
+ 
++    //Allows for textures located outside the vanilla horse armor folder
++    private HorseArmorType(String textureLocation, String hash, int armorStrengthIn)
```
Please keep all new methods in a `/********* FORGE *********/` section at the bottom of the file.

mezz commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,19 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
+     * 
+     * @param name - The name the new enum constant should have
+     * @param textureLocation - The path to the texture for this armor type. Formatted as domain:path.
+     * @param hash - A short unique 3 letter name for the armor type 
```
In vanilla, the hash is used for the horse texture prefix. 
Is that still relevant here, or should we hard-code it to `""` instead of making it a parameter here?

Daomephsta replied:
> Is that still relevant here

Currently, yes.

I've just realised that there is a significant issue with the hashes. Each HorseArmorType **must** have a unique hash. If two types have the same hash, the wrong texture will be rendered in certain cases.  

An example:  
There are two horses, A & B, both have the **same base coat and markings**. There are also two horse armors, Y & Z, both have the **same hash**. A wears armor Y, B wears armor Z. The position of the two horses doesn't matter, as long as the player sees them in the same Minecraft session. If A is rendered first, both armors will use armor Y's texture. If B is rendered first, both armors will use armor Z's texture. This is due to MC caching the layered texture for a particular combination of coat, markings and armor in a map with the concatenated hashes as the key.

I can think of two solutions to this:  
1. Prepend the modid to the hash. 
2. Generate hashes that are three letter alphabetic sequences. So the first armor type registered would get the hash 'aaa', the second 'aab', the third 'aac', etc. This would have a negligible memory and CPU cost. However it would limit the number of possible horse armor type to 26^3(17,576), that should be more than enough though. The hashes don't need to be persistent as the texture cache isn't persistent.

1 is simple, however I'd prefer 2 as the hash is really an internal implementation detail that a modder doesn't need to deal with. Since we can generate it, I think we should. I also prefer 2 as it keeps the hashes at their current length. I wrote a string hashcode [test program](https://gist.github.com/Leviathan143/d7c0da865dbbbf11615a2f2357d49d3f) and the hashcode computation times are extremely fast (The data and best/worst case scenarios are included in the linked gist), however I assume the hashes were kept short for some reason and using 2 means we don't have to worry about what that reason is.

mezz replied:
- If that hash must match on client and server or needs to be persistent across game loads:
   You would probably have to use a registry. I don't think enough mods will be adding horse armor to justify an advanced solution. Just have mods hard-code their values and negotiate collisions manually.
- If the hash is only used on the client:
   I think you could simply generate the hashes automatically starting with "aaa", "aab", etc. Just make sure they don't collide with vanilla's.

Daomephsta replied:
Thanks for the feedback.
Hashes are client-side only and do not need to be persistent, so I'll be going with solution 2.
I assume the hash generation method would go in ForgeHooksClient?

mezz replied:
Sounds good. There's not really an ideal place for it so `ForgeHooksClient` should be fine.

mezz reviewed (changes requested):
Looks pretty good but needs some fixes before review.
Also please make the textures anything besides Pink and Black, which are normally used to indicate something is broken. Green and Orange is fine.

leviathan143 committed:
Move new methods and constructor into Forge section

leviathan143 committed:
Implement automatic hash generation for horse armor to prevent hash collisions

leviathan143 committed:
Make textures look less like broken texture

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,65 @@
+         this.field_188587_f = "";
+     }
+-
++    
```
This line should not be changed

mezz commented on patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:
```diff
@@ -0,0 +1,64 @@
+             this.func_110230_cF();
+         }
++        ItemStack armor = this.field_110296_bG.func_70301_a(1);
++        if(armor.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor) ((net.minecraftforge.common.IHorseArmor) armor.func_77973_b()).onHorseArmorTick(field_70170_p, this, armor);
```
`if (` should have a space between `if` and `(`

mezz commented on patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:
```diff
@@ -0,0 +1,64 @@
+     {
+-        return HorseArmorType.func_188575_a(((Integer)this.field_70180_af.func_187225_a(field_184791_bI)).intValue());
++        HorseArmorType armor = HorseArmorType.getByName(this.field_70180_af.func_187225_a(HORSE_ARMOR_STRING)); //First check the Forge armor DataParameter
++        if(armor == HorseArmorType.NONE) armor = HorseArmorType.func_188575_a(this.field_70180_af.func_187225_a(field_184791_bI)); //If the Forge armor DataParameter returns NONE, fallback to the vanilla armor DataParameter. This is necessary to prevent issues with Forge clients connected to vanilla servers.
```
`if (` should have a space between `if` and `(`

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,65 @@
+     public static HorseArmorType func_188580_a(ItemStack p_188580_0_)
+     {
+-        return p_188580_0_.func_190926_b() ? NONE : func_188576_a(p_188580_0_.func_77973_b());
++        if(p_188580_0_.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor)
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+    /**For internal use only**/
+    public static String generateHorseArmorHash()
+    {
+        if(lastCharIndices[0] == -1)
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+        {
+            hashChars[c] = ALPHABET[lastCharIndices[c]];
+        }
+        if(lastCharIndices[2]++ >= ALPHABET.length - 1)
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+            if(lastCharIndices[1]++ >= ALPHABET.length - 1)
+            {
+                lastCharIndices[1] = 0;
+                if(lastCharIndices[0]++ >= ALPHABET.length - 1)
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+        if(lastCharIndices[2]++ >= ALPHABET.length - 1)
+        {
+            lastCharIndices[2] = 0;
+            if(lastCharIndices[1]++ >= ALPHABET.length - 1)
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+            throw new RuntimeException("No available horse armor hash.");
+        }
+        char[] hashChars = new char[3];
+        for(int c = 0; c < 3; c++)
```
`for (` should have a space between `for` and `(`

mezz commented on src/test/java/net/minecraftforge/test/EnumHelperTest.java:
```diff
@@ -90,6 +91,10 @@ public void testEnumHelperMethodsMatchEnumConstructors()
                     {
                         filter = true; //We don't want people using this method.
                     }
+                    if(returnType == HorseArmorType.class && (declaredConstructor.getParameterTypes().length == 3 || declaredConstructor.getParameterTypes()[2] == int.class))
```
`if (` should have a space between `if` and `(`

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -714,4 +714,37 @@ public static ScreenshotEvent onScreenshot(BufferedImage image, File screenshotF
+            }
+        }
+        String result = String.valueOf(hashChars);
+        //Make sure we don't return a hash used by vanilla. We can't compare to HorseArmorType#getHash() because that method does not exist on the DS
```
You should be able to remove `@SideOnly(Side.CLIENT)` from `getHash` without breaking anything since nobody can extend it.

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,65 @@
+         }
+     }
+ 
++    @Deprecated
```
put a comment saying what method to use instead:
`@Deprecated // Forge: use ___ method instead`

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,65 @@
++        return p_188580_0_.func_77973_b() == Items.field_151138_bX ? net.minecraft.entity.passive.HorseArmorType.IRON : p_188580_0_.func_77973_b() == Items.field_151136_bY ? net.minecraft.entity.passive.HorseArmorType.GOLD : p_188580_0_.func_77973_b() == Items.field_151125_bZ ? net.minecraft.entity.passive.HorseArmorType.DIAMOND : net.minecraft.entity.passive.HorseArmorType.NONE;
+     }
+ 
++    @Deprecated
```
put a comment saying what method to use instead:
`@Deprecated // Forge: use ___ method instead`

mezz reviewed (changes requested):
Looks pretty good, mostly small style issues.

leviathan143 committed:
Fix style issues

leviathan143 committed:
Remove @SideOnly annotation from HorseArmorType#getHash()

howtonotwin commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,64 @@
+         return this.field_188586_e;
+     }
+ 
++    @Deprecated
```
You should comment why.

howtonotwin commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,22 @@
+public interface IHorseArmor 
+{
+    /** 
+     * @param stack - The armor stack
```
Incorrect Javadoc formatting. After `@param`, there should be no dashes or any other kind of separator, because that's what the Javadoc compiler is for. Also, if the first phrase of the string is not a full sentence, then it is not capitalized, and if that first non-sentence phrase is the only phrase, it is not ended with a period. If there is another phrase, then there is always a period, and every phrase that is not the first should be a sentence. [Reference (`@param`)](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@param)

This line should be:

    @param stack the armor stack

howtonotwin commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface IHorseArmor 
+{
+    /** 
```
Please write a sentence or two of description here. Even if it is completely redundant with `@return`, it is still good style to have it. [Reference (`@return`)](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return)

howtonotwin commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,22 @@
+{
+    /** 
+     * @param stack - The armor stack
+     * @return The HorseArmorType of the armor stack 
```
`@return` follows the same casing and punctuation conventions as `@param`. Also, please wrap any Java identifiers in `{@code}`. [Reference (style guide)](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#styleguide) [Reference (`@return`)](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return)

howtonotwin commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,22 @@
+     */
+    public HorseArmorType getHorseArmorType(ItemStack stack);
+    
+    /**
```
Add a description, like above

howtonotwin commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * @param world - The world the horse is in
+     * @param horse - The horse wearing the armor
+     * @param itemStack - the armor itemstack
```
Fix the `@param`s, as above. Also, it may be a good idea to rename `itemStack` -> `armor`.

howtonotwin commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,18 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
         return addEnum(EnumRarity.class, name, color, displayName);
     }
+    /**
+     * 
```
Add a description, as above.

howtonotwin commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,18 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
+     * 
+     * @param name - The name the new enum constant should have
+     * @param textureLocation - The path to the texture for this armor type. Formatted as domain:path.
+     * @param armorStrength - How much protection this armor type should give
```
Fix the `@param`s, as above.

howtonotwin commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,18 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
+    /**
+     * 
+     * @param name - The name the new enum constant should have
+     * @param textureLocation - The path to the texture for this armor type. Formatted as domain:path.
```
What path is the `path` relative to? Does `a:b` mean `assets/a/textures/b.png`? `assets/a/textures/horse_armor/b.png`? `assets/a/b.png`?

howtonotwin commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,18 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
     }
+    /**
+     * 
+     * @param name - The name the new enum constant should have
```
Don't say "the name the new enum constant should have," say "the name of the new {<span>@</span>code HorseArmorType}".

howtonotwin commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -141,6 +143,18 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
+     * @param name - The name the new enum constant should have
+     * @param textureLocation - The path to the texture for this armor type. Formatted as domain:path.
+     * @param armorStrength - How much protection this armor type should give
+     * @return The created HorseArmorType enum constant
```
The wording is a bit awkward here. Maybe just say "The new {<span>@</span>code HorseArmorType}". Also, this method is `@Nullable`, which seems like a special case and should be documented in `@return`. [Reference (`@return`)](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return)

howtonotwin commented on src/test/java/net/minecraftforge/debug/HorseArmorTest.java:
```diff
@@ -0,0 +1,73 @@
+    public static final boolean ENABLED = false;
+    
+    public static HorseArmorType testArmorType;
+    @ObjectHolder(MODID + ":test_armor")
```
There is no point in `@ObjectHolder`ing your own objects. Just instantiate it and and call the setters here.

covers1624 replied:
Most of the other debug mods do.

marvin-roesch replied:
It's also the recommended way to do this, test mods should follow standard practices.

howtonotwin replied:
@PaleoCrafter, by "it," do you mean with a holder or without? If the former, I believe Lex once said not to use the annotation like this once on IRC. That was a long time ago, though. Stances may have changed.

marvin-roesch replied:
I mean with the holder. Although in reality it shouldn't be the case that test mod blocks/items get overridden, they should reflect best practices. The only minor nitpick I'd have is that the annotation is used on a field level rather than a class one, which is preferred since it requires fewer "magic strings".

bs2609 replied:
If you put the annotation on the class, it'll take all `public static final` fields in that class to be object holders. Which I believe will cause a fail here, as `String` isn't a valid object holder type.

bs2609 replied:
In this instance, the mod id isn't required for the annotation, as it's in the `@Mod` class.

howtonotwin commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,64 @@
+         }
+     }
+ 
++    @Deprecated //Forge: Use ItemStack sensitive overload
```
I believe it's our convention to have spaces after `//`, but meh.

howtonotwin commented on src/test/java/net/minecraftforge/debug/HorseArmorTest.java:
```diff
@@ -0,0 +1,73 @@
+        } 
+        
+        @SubscribeEvent
+        public static void registerModels(ModelRegistryEvent event) 
```
This will break a server. Move `registerItems` into the `@Mod` class, add `@EventBusSubsciber` to the `@Mod` class, and add `value = Side.CLIENT` to this nested class's `@EventBusSubscriber`.

liach replied:
No, this will not. That event is only posted on the client.

marvin-roesch replied:
It will fail to load the class on a server, due to the method being present with that signature. That will cause a cascade of loaded classes, some of which are client-only. See #4225 for similar examples.

liach replied:
A `@SideOnly(Side.CLIENT)` is enough, not to be too complicated like @howtonotwin suggests.

marvin-roesch replied:
See Lex' comment on that PR (#4225), `@SideOnly` is not an option here or, as a matter of fact, in any mod.

liach replied:
Bus subscriber is worse. You should avoid statics since java is oop

liach replied:
And `@SideOnly` is a mean to control code flow. I don't see why it is strongly discouraged, as it is not told anywhere.

howtonotwin reviewed (changes requested):
Looks good; just some nitpicks.

leviathan143 committed:
Improve Javadocs

leviathan143 committed:
Prevent crash on DS

leviathan143 committed:
Add default modifier to IHorseArmor#onHorseArmorTick

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -12,7 +12,7 @@
      }
  
-+    @Deprecated
++    @Deprecated /**Forge: Use getByName. Ordinals of mod-added enum constants are dependent on load order, unlike names.**/
```
Use `/* */` not `/** **/`

liach commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -8,15 +8,18 @@
+     * @param armor the armor itemstack
      */
-    public void onHorseArmorTick(World world, EntityHorse horse, ItemStack itemStack);
+    public default void onHorseArmorTick(World world, EntityHorse horse, ItemStack armor) {}
```
no need for public for both

liach commented on src/test/java/net/minecraftforge/debug/HorseArmorTest.java:
```diff
@@ -18,15 +18,17 @@
 {
     public static final String MODID = "horse_armor_test";
-    public static final boolean ENABLED = false;
+    public static final boolean ENABLED = true;
```
lolol

leviathan143 committed:
Make test mod disabled by default

leviathan143 committed:
Remove redundant public access modifiers from IHorseArmor methods

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,64 @@
+         return func_188576_a(p_188577_0_) != NONE;
+     }
++    
++    /********* FORGE *********/
```
Dangling javadocs

Daomephsta replied:
That's what mezz told me to put there. There's no standard for forge section delimiters, so I'm sticking with that.

bs2609 replied:
```java
/* ======================================== FORGE START ===================================== */
```
seems pretty standard.

If you want to stick with that, please do use spaces inside the `/*` and `*/` parts to delimit the comment markers from the actual comment. As you have it now, it's an erroneous Javadoc comment.

Daomephsta replied:
Valid point

LexManos replied:
I usually just hold down = until it fills up 1/2 my screen :P it's intended to be fairly big and obvious.

liach commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,64 @@
++    {
++        this.field_188588_g = armorStrengthIn;
++        this.field_188586_e = textureLocation;
++        this.field_188587_f = net.minecraftforge.client.ForgeHooksClient.generateHorseArmorHash();
```
Although this is only used on the client, this should not be in `ForgeHooksClient` for its existence on the server side.

leviathan143 committed:
Move hash generator to ForgeHooks as the hash is now accessible on both server and client.

leviathan143 committed:
Make vanilla hash check more robust

leviathan143 committed:
Make Forge section delimiters consistent with the majority of Forge.

mezz commented:
Looks good, please resolve the merge conflicts.

leviathan143 committed:
Merge branch '1.12.x' into horse-armour-1.12

bs2609 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1347,6 +1348,45 @@ private static boolean loadAdvancements(Map<ResourceLocation, Advancement.Builde
+            }
+        }
+        return result;
+      
```
`}`!
Looks like broken merge?

leviathan143 committed:
Add missing }

mezz commented:
Tested and looks good to me. I wasn't actually able to see the horse armor though, the glow effect is too strong to see anything.

Daomephsta commented:
@mezz That's odd. It looks fine for me. You weren't looking from a distance were you? The glow effect maintains a constant thickness, so at long range it does obscure the entity.
![2017-10-02_08 21 08](https://user-images.githubusercontent.com/10936237/31058117-cfb6ba0c-a74a-11e7-8d7f-e04886773155.png)

mezz commented:
My resources weren't loaded properly. I tested again and it looks fine.

LexManos commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,25 @@
+import net.minecraft.world.World;
+
+public interface IHorseArmor 
+{
```
Why is this an interface instead of just getHorseArmorType on item returning null?

Daomephsta replied:
Initially it was the latter, as my understanding is that Forge has been against using interfaces in this scenario (at least prior to 1.12 requiring Java 8) for 2 reasons: 
1. Adding new methods to an interface breaks binary compat.
2. Calling the interface methods requires an instanceof check and a cast, which is messy.  

Are these reasons correct, and are there any more?

1.12 requiring Java 8 solves the first problem with the default modifier. The second problem is not much of an issue in this case as the methods are called in only two places. Since I thought Forge was against using interfaces in this scenario, I didn't use one. Then @liach implied that Forge convention had changed in that regard, so I changed it to an interface because I find that cleaner than shoving more methods in Item. If Forge convention has not changed, then I will change it back.

liach replied:
I suggested for interfaces like `IPlantable` and `ISpecialArmor` exist. Move it into item class then.

Daomephsta replied:
@LexManos It's been two weeks since I replied to your question about why I used an interface. Can I please get an answer?

LexManos commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1347,6 +1348,46 @@ private static boolean loadAdvancements(Map<ResourceLocation, Advancement.Builde
+    private static final char[] ALPHABET = "abcdefghijklmnopqrstuvwxyz".toCharArray();
+    private static int[] lastCharIndices = new int[3];
+    /**For internal use only**/
+    public static String generateHorseArmorHash()
```
Why make this complicated 'hash' {more if a suffix}  Why not just call make the armor return the full texture path?

KingLemming replied:
Agreed. Resource locations ftw.

Daomephsta replied:
It's part of the vanilla implementation. The hash is used to form a short unique ID for each combination of horse armour, base coat and markings. This ID is used to construct a ResourceLocation, which is used as the ID in the TextureManager for the LayeredTexture created from that combination.

I'm not sure why Mojang has done this. Replacing the hash with the full texture location of the armour texture would greatly lengthen the ID, which would slow down its hashcode function and hence make the texture retrieval slower. However `String#hashCode()` is damn fast, run times are measured in nanoseconds. 

In the end I opted to keep the hash system, because Mojang has gone to some trouble to make it, so they must have a good reason. Making an automatic hash generator was easy, so I did that rather than removing the system and worrying about why Mojang made it.

LexManos replied:
Ya, don't use small hashes. Its just super complex for no reason. I can ask around but im pretty sure the reason Mojang did it was because "Meh" Performance wise it isnt an issue.

Daomephsta replied:
Okay, that simplifies things a lot.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,65 @@
++    @Deprecated //Forge: Use getByItemStack
+     public static HorseArmorType func_188576_a(Item p_188576_0_)
+     {
+         if (p_188576_0_ == Items.field_151138_bX)
```
Why not just Item.getHorseArmorType(stack){ return HorseArmorType.getByItem(stack.getItem()); }
Less code changes and better compatibility with future changes.

LexManos reviewed (commented):
Ya, don't use an interface.

leviathan143 committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into horse-armour-1.12

Daomephsta commented:
The requested changes to this PR are almost complete. I'm waiting until I've done my last exam, before I finish and commit the changes.

leviathan143 committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into horse-armour-1.12

leviathan143 committed:
Remove IHorseArmor interface in favour of patching methods into Item.
Sync horse armour stack using data parameters so properties(texture etc.) of horse armour can be NBT/metadata/whatever based.

Daomephsta commented:
@mezz I've made all the requested changes. Can you take another look at this please?
Note that if you want to test vanilla clients connecting to Forge servers, you'll need to add `acceptableRemoteVersions = "*"` to the following test mods:  

- CheckSpawnTest  
- ChunkCapabilityPollutionTest  
- DebugSearchTabs  
- ItemModelGenerationTest  
- ReachDistanceAttributeTest  
- RegistryOverrideTest  
- SlipperinessTest  
- TransparentFastTESRTest  
- SelectorHandlerTest  

I'll be making a PR shortly to fix these test mods so that they will accept vanilla clients, as they should when disabled. This is an issue with the listed test mods, not this PR.

mezz commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,61 @@
+         return this.ordinal();
+     }
+ 
+-    @SideOnly(Side.CLIENT)
```
Since you are not using `getHash()` any more, you should leave the `@SideOnly(Side.CLIENT)` there, and consider deprecating it.

mezz commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -55,6 +55,7 @@
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.item.EntityMinecartContainer;
+import net.minecraft.entity.passive.HorseArmorType;
```
Please remove changes to this file, there is nothing here any more.

mezz reviewed (changes requested):
Looking good, just missed some things for cleanup from the latest changes.

leviathan143 committed:
Clean up loose ends from previous commit

leviathan143 committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into horse-armour-1.12
